catroid/src/main/java/org/catrobat/catroid/web/ProjectDownloader.kt

```diff
@@ -0,0 +1,183 @@
+import android.os.Handler
+import android.os.ResultReceiver
+import android.support.annotation.VisibleForTesting
+import android.support.v7.app.AppCompatActivity
+import android.util.Log
+import org.catrobat.catroid.R.string.error_project_already_in_queue
+import org.catrobat.catroid.common.FlavoredConstants.DEFAULT_ROOT_DIRECTORY
+import org.catrobat.catroid.scratchconverter.Client.ProjectDownloadCallback
+import org.catrobat.catroid.transfers.project.ProjectDownloadService
+import org.catrobat.catroid.transfers.project.ProjectDownloadService.Companion.ERROR_CODE
+import org.catrobat.catroid.transfers.project.ProjectDownloadService.Companion.SUCCESSFUL_CODE
+import org.catrobat.catroid.transfers.project.ProjectDownloadService.Companion.UPDATE_PROGRESS_CODE
+import org.catrobat.catroid.transfers.project.ProjectDownloadService.Companion.UPDATE_PROGRESS_EXTRA
+import org.catrobat.catroid.ui.recyclerview.dialog.ReplaceExistingProjectDialogFragment
+import org.catrobat.catroid.utils.ToastUtil
+import org.catrobat.catroid.utils.notifications.StatusBarNotificationManager
+import java.io.File
+import java.io.Serializable
+import java.io.UnsupportedEncodingException
+import java.lang.ref.WeakReference
+import java.net.URLDecoder
+import java.util.Collections
+import java.util.HashSet
+import java.util.Locale
+
+class ProjectDownloader(
+    private val queue: ProjectDownloadQueue,
+    private val url: String,
+    callback: ProjectDownloadCallback?
+) : Serializable {
+    private val callbackWeakReference = WeakReference<ProjectDownloadCallback>(callback)
+
+    companion object {
+        private const val serialVersionUID = 42L
+        private const val FILENAME_TAG = "fname="
+        private val TAG = ProjectDownloader::class.java.simpleName
+
+        fun getProjectNameFromUrl(url: String): String? {
+            val projectNameIndex = url.lastIndexOf(FILENAME_TAG) + FILENAME_TAG.length
+            val projectNameUTF8 = url.substring(projectNameIndex)
+            return try {
+                URLDecoder.decode(projectNameUTF8, "UTF-8")
+            } catch (e: UnsupportedEncodingException) {
+                Log.e(TAG, "Could not decode project name: $projectNameUTF8", e)
+                null
+            }
+        }
+    }
+
+    fun download(activity: AppCompatActivity) {
+        val projectName = getProjectNameFromUrl(url) ?: return
+
+        if (queue.projectNameNotAvailable(projectName)) {
+            val dialog = ReplaceExistingProjectDialogFragment.newInstance(projectName, this)
+            dialog.show(activity.supportFragmentManager, ReplaceExistingProjectDialogFragment.TAG)
+        } else {
+            downloadOverwriteExistingProject(activity, projectName, false)
+        }
+    }
+
+    fun downloadOverwriteExistingProject(
+        context: Context,
+        projectName: String,
+        renameProject: Boolean
+    ) {
+        synchronized(queue) {
+            if (queue.alreadyInQueue(projectName)) {
```

Comment:
check the escaping of special characters!!!
pocket code escapes spec. chars as far as i know,.so will share for display in the url (eg. "?",""",etc)

Reply:
The project name in the the URL is utf8 encoded and is decoded before putting it into the queue.
To eliminate the problem that (maybe, in some way) the project name in the XML and the name in the URL are not equal, I'll remove the boolean flag (renameProjectAfter download), so that the project name will be always renamed after download.